src/registration.ts

```diff
@@ -1,55 +1,64 @@
 import { number, object, string, ValidationError } from "yup";
+import { map, skip, tap } from "rxjs/operators";
+import { combineLatest } from "rxjs";
+import { createPipelineFromInput$ } from "./helpers/pipeline.helper";
+import { createDefaultState, FormError, RegistrationForm, RegistrationFormState } from "./models/form.registration";
 import "./assets/styles/styles.scss";
 
+const formErrorsEl = document.querySelector(".form-errors");
+const usernameElement = document.querySelector("#username") as HTMLInputElement;
+const emailElement = document.querySelector("#email") as HTMLInputElement;
+const passwordElement = document.querySelector("#password") as HTMLInputElement;
+const ageElement = document.querySelector("#age") as HTMLInputElement;
+
 document.addEventListener("DOMContentLoaded", init);
 
-function init() {
-  const registrationForm = {
-    username: "",
-    password: "",
-    email: "",
-    age: ""
-  };
+const registrationValidationSchema = object().shape({
+  username: string().required("Username is required").min(4, "Should be more than 4 symbols"),
+  email: string().required("Email is required").email("Email is required"),
+  password: string().required("Password is required").min(4, "Password should be more than 4 symbols"),
+  age: number().required("Number is required").min(0)
+});
+const formData: RegistrationForm = {
+  username: usernameElement.value,
+  email: emailElement.value,
+  password: passwordElement.value,
+  age: Number(ageElement.value),
+}
+const formState: RegistrationFormState = createDefaultState();
 
-  // TODO:  On change of input fields show validation errors, form state - valid/invalid and formData
-  // TODO: create fromEvent listeners for username password etc., use mapToTarget, startWith and map to value => ({ [key]: value })
-  // TODO: then track form changes (like in balance.ts) and validate form, show errors, current form data on UI
-  // TIPS: you can use Object.keys, fromEvent, map, startWith, combineLatest, js reduce operator etc.
+function init() {
```

Comment:
const mapToTarget = map((e: Event) => (e.target as HTMLInputElement).value);
  const registrationForm = {
    username: "",
    password: "",
    email: "",
    age: ""
  };

  const combinedSourceArr$ = Object.keys(registrationForm).map(key => {
    return fromEvent(document.querySelector(`#${key}`), "input").pipe(
      mapToTarget,
      startWith((document.querySelector(`#${key}`) as HTMLInputElement).value),
      **map(value => ({ [key]: value }))**
    );
  });

/// creates
{
 username: value,
 password: value,
....
}

  combineLatest(...combinedSourceArr$)
    .pipe(
      **_map(errorList => errorList.reduce((acc, val) => ({...acc, ...val}), {}))_**
    )
    .subscribe(data => console.log('return', data));



_Consider the following code - 
I guess we can do a little bit generic._